### PR TITLE
[kdeconnect] Improvements

### PR DESCRIPTION
**Pull request by raffaem**
- Support multiple devices
- Filter out non-regular files

### CODE REVIEW

File: plugins/kdeconnect [hunk 1]
```diff
@@ -26,12 +26,36 @@ send() {
...
+# Select paired device
+names=$(kdeconnect-cli --list-available --name-only 2>/dev/null)
+if [ -z "$names" ]; then
+	note "No devices paired and available"
 	exit
 fi
 
+ndevs=$(printf "%s" "$names" | awk 'END{print NR}')
+if [ "$ndevs" -eq 1 ]; then
+	devname=$(printf "%s" "$names" | awk 'NR==1')
```
Comment (N-R-K) [inline plugins/kdeconnect:38]: If there's just one name, why call awk on it?

File: plugins/kdeconnect [hunk 2]
```diff
@@ -26,12 +26,36 @@ send() {
...
+else
+	# Append a newline at the end of `names` to read the last device
+	# Note that command substitution removes trailing newlines
+	# see: https://stackoverflow.com/a/15184414/1719931
+	names=$(printf "%s\nx" "$names");
+	names=${names%x}
+	i=0
+	printf "%s" "$names" | while IFS=$(printf "\n") read -r name; do
```
Comment (N-R-K) [inline plugins/kdeconnect:46]: This loop can probably be replaced by something like this ` printf "%s" "$names" | awk '{ print NR ". " $0 }'`.

File: plugins/kdeconnect [hunk 3]
```diff
@@ -26,12 +26,36 @@ send() {
...
+		printf "%d. %s\n" "$i" "$name"
+		i=$((i+1))
+	done
+	printf "Pick a device: "
+	read -r pick
+	if [ -n "$pick" ] && [ "$pick" -eq "$pick" ]; then
+		pick=$((pick+1))
+		devname=$(printf '%s' "$names" | awk "NR==$pick")
+		printf "You have selected: %s\n" "$devname"
```
Comment (N-R-K) [inline plugins/kdeconnect:55]: Don't think this is necessary. The user can figure out what he selected by looking at the list.

**Review by N-R-K — CHANGES_REQUESTED** [review]

File: plugins/kdeconnect [hunk 4]
```diff
@@ -22,16 +22,29 @@ note() {
 }
 
 send() {
-	xargs -0 -I{} kdeconnect-cli --name "$devname" --share {}
+	xargs -0 -I{} sh -c 'if [ ! -d "{}" ]; then printf "%s\0" "{}"; fi;' |  xargs -0 -I{} kdeconnect-cli --name "$devname" --share {}
```
Comment (N-R-K) [inline plugins/kdeconnect:25]: What's this about?

Comment (raffaem) [reply plugins/kdeconnect:25]: if you pass a directory, kndeconnect creates an empty file on the device with the name of the directory. Basically it doesn't support directory transfers.

Comment (N-R-K) [reply plugins/kdeconnect:25]: As far as I see, this simply filters out directories, right? It doesn't recursively send all the files inside the dir?

If that's the case, then just leave it out. The user can simply not select a directory. And if he does select one, it's better to see the empty file than to silently discard his selection.

Comment (raffaem) [reply plugins/kdeconnect:25]: Yes it just filters out directory.

1. But I think we will need some kind of checks. What happens if you send over a block device?
2. In general, how can I have xargs call a user-defined function? On SO [I found I can export the function](https://stackoverflow.com/a/11003457/1719931), but `shellcheck` complains that is not POSIX sh-compatible.

Comment (raffaem) [reply plugins/kdeconnect]: it hangs

Comment (N-R-K) [reply plugins/kdeconnect]: In that case, filtering out only the files seems like a good idea. Does this work?

```diff
diff --git a/plugins/kdeconnect b/plugins/kdeconnect
index 0d3ef8b7..209bf137 100755
--- a/plugins/kdeconnect
+++ b/plugins/kdeconnect
@@ -21,8 +21,18 @@ note() {
 	fi
 }
 
+# kdeconnect doesn't cope with non-files
+filter_files() {
+	xargs -0 -I{} sh -c \
+		'if [ -f "{}" ]; then'\
+		'	printf "%s\0" "{}";'\
+		'else'\
+		'	printf "Error: not a file: %s\n" "{}" >&2;'\
+		'fi;'
+}
+
 send() {
-	xargs -0 -I{} kdeconnect-cli --name "$devname" --share {}
+	filter_files | xargs -0 -I{} kdeconnect-cli --name "$devname" --share {}
 	note "Files sent"
 }
 
```

Comment (raffaem) [reply plugins/kdeconnect]: No it says:
```
Pick a device: 1
	printf "%s\0" "src";: -c: line 2: syntax error: unexpected end of file
	printf "%s\0" "CHANGELOG";: -c: line 2: syntax error: unexpected end of file
	printf "%s\0" "LICENSE";: -c: line 2: syntax error: unexpected end of file
```

but I made it to work by removing the `\` characters :)

Comment (raffaem) [reply plugins/kdeconnect]: How can we call:

```
note "Error: not a file:
```

instead of:

```
printf "Error: not a file:
```
?

Comment (N-R-K) [reply plugins/kdeconnect]: Don't think that's possible without doing something dirty. So I'd leave it be.

**Review by N-R-K — APPROVED** [review]
@jarun This should be good to merge.

**Comment by jarun** [comment]
Thanks guys!

**Pull request merged**